main.js

```diff
@@ -0,0 +1,1512 @@
+    iconText.textContent = "add";
+  } else {
+    iconText.textContent = "remove";
+  }
+}
+function subtaskCount() {
+  tasks.forEach((task) => {
+    const taskId = task.id;
+    const totalCount = task.subtasks.length;
+    const completedCount = task.subtasks.filter(
+      (subtask) => subtask.complete === true
+    ).length;
+    const inCompletedCount = totalCount - completedCount;
+    renderSubtaskCount(taskId, inCompletedCount, totalCount);
+    const completionRate = (completedCount / totalCount) * 100 || 0;
+    renderSubtaskProgress(taskId, completionRate);
+  });
+}
+function renderSubtaskProgress(Id, rate) {
+  const taskCardLi = document.querySelector(`[data-task-id="${Id}"]`);
+  const progressFill = taskCardLi.querySelector(`[data-progress-fill]`);
+  const progressValue = taskCardLi.querySelector(`[data-progress-value]`);
+  const roundedRate = Math.round(rate) + "%";
+  const percentage = rate + "%";
+  progressValue.innerText = roundedRate;
+  progressFill.style.width = percentage;
+}
+function renderSubtaskCount(Id, inCompletedCount, totalCount) {
+  const taskCardLi = document.querySelector(`[data-task-id="${Id}"]`);
+  const subtaskText = taskCardLi.querySelector(`[data-subtasks-left]`);
+  if (totalCount === 0) {
+    subtaskText.innerText = `No tasks yet`;
+  } else if (inCompletedCount === 0 && totalCount > 0) {
+    subtaskText.innerText = `Completed!`;
+  } else {
+    subtaskText.innerText = `${inCompletedCount} Left`;
+  }
+}
+
+const getData = function () {
+  const taskId = Date.now().toString();
+  const taskName = newTaskNameInput.value;
+  const taskStarts = newTaskStartInput.value;
+  const taskEnds = newTaskDeadlineInput.value;
+  const taskDescription = newTaskDescriptionInput.value;
+  pushData(taskId, taskName, taskStarts, taskEnds, taskDescription);
+};
+
+function pushData(taskId, taskName, taskStarts, taskEnds, taskDescription) {
+  //should return some error to
+  if (taskName == null || taskName === "") return;
+  if (validateTaskDate() !== "Good to go") return;
+  const task = createTask(
+    taskId,
+    taskName,
+    taskStarts,
+    taskEnds,
+    taskDescription
+  );
+  tasks.push(task);
+  makeSubtaskArray(taskId);
+  save();
+  renderAndReOpen();
+  resetForm();
+}
+
+function resetForm() {
+  newTaskNameInput.value = "";
+  newTaskDescriptionInput.value = "";
+  clearElement(newTaskSubtaskContainer);
+  resetNewtaskIcons();
+}
+
+function resetNewtaskIcons() {
+  const switchIcons = newTaskPanel.querySelectorAll(".toggle-switch");
+  switchIcons.forEach((icon) => {
+    if (icon.textContent == "toggle_on") {
+      toggleSwitchIcon(icon);
+      toggleHideShow(icon.parentElement.nextElementSibling);
+    }
+  });
+}
+
+function makeSubtaskArray(Id) {
+  let selectAllSubtaskForms = document.querySelectorAll(".new-subtask");
+  let subtaskFormsArray = [...selectAllSubtaskForms];
+  let notEmpty = subtaskFormsArray.filter((e) => e.value.trim().length > 0);
```

Comment:
`filter` method already returns with a new array with new reference so there is no need to use spread operator here `[...selectAllSubtaskForms]` if the intention is copying element as a new array.

Reply:
That is because, document.querySelectorAll() method currently returns NodeLists and NodeLists currently cannot be accessed by array methods like filter(), forEach(), some() etc. To be able to access and use its values like this it needs to be converted to an array. So copying it to a new array with spread operator was the best chance(And probably still is 😅).